### ISSUE TITLE
Perf testing: write throughput bottlenecks

### ISSUE THREAD

**Issue by tleyden** [couchbase/sync_gateway#853]
In perf runs which have [extra view queries triggered](https://github.com/couchbase/sync_gateway/issues/825)

In [perf run 921](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/921/):
- 1k pushers
- 1k pullers
- Couchbase Server 2.5

total doc pushed = **67K**

In [perf run 924](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/924/):
- 5k pushers
- 5k pullers
- Couchbase Server 2.5

total doc pushed = **46K**  (expected: this number should be much higher than the 1k/1k case)

Digging into this, it appears that there is a lot of write backpressure, and so gateload's writes are taking a long time.
- Perf run 921 PostBulkDocs P95: **76 seconds**
- Perf run 924 PostBulkDocs P95: **1189 seconds**

**Comment by tleyden** [comment]
For comparison purposes, [perf run 898](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/898/):
- 5k pushers
- 8k pullers
- Couchbase Server 4.0.0-1817

total docs pushed = **480K per gateload**

This run is doing less writes from Sync Gw -> CBS for each gateload write, because with extra view queries triggered as in [perf run 921](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/921/) and  [perf run 924](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/924/):
- every write ends up spawning multiple additional writes, to zero out the user doc
- (maybe) requires an additional INCR as well

**Comment by adamcfraser** [comment]
May be useful to drill into the cb/ops numbers in the Sync Gateway expvars for perf runs 921 and 924.   924 doesn't actually show a higher number of ops for any operation, except for casNext.  The mean time for each of the ops is quite a bit higher in 924, however.

**Comment by adamcfraser** [comment]
Another point for investigation - the cb/pools expvars.  The 924 count is only moderately higher (~1.2x), but the mean times are more than 10x higher.

**Comment by tleyden** [comment]
TODO: retest with later go-couchbase version that has a higher connection pool setting to see if this makes any difference.

**Comment by tleyden** [comment]
Extracted from https://github.com/couchbase/sync_gateway/issues/781

Action item:
- Look at total number of cb ops (add together) in sync gw expvars
- Get exact total time of run
- Get total ops/sec
- Verify size of connection pool (12?)

**Comment by tleyden** [comment]
Notes: SG already has a config parameter that sets go-couchbase PoolSize.  https://github.com/couchbase/sync_gateway/blob/master/src/github.com/couchbase/sync_gateway/rest/config.go#L67

We're currently defaulting to 16, with zero overflow.

If we don't set anything in our config, we would have been overriding go-couchbase'es new default value without us realizing it.

**Comment by tleyden** [comment]
Increasing the number of Couchbase connections to 250 didn't have any effect.
- 1k/1k: [perf run 935](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/935/) -- total docs pushed: **69K**
- 5k/5k: [perf run 936](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/936/) -- total docs pushed: **50K**

The way to increase the connections is to add this override: **gateway.conn_db.250**

I did verify that the sync gateway config contained `maxCouchbaseConnections: 250,`, but did not do any verification above that.

**Comment by tleyden** [comment]
I added println tracing/instrumentation to the `feature/issue_853` branch which emits logs whenever a few instrumented functions are taking longer than 1 second.  I'm trying to figure out where the writes are bottlenecked.  In [perf run 939](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/939/) I ran this with 
- 5k pushers
- 5k pullers
- Couchbase Server 2.5

and saw a lot of cases in the logs where the instrumented functions were taking much longer than 1s:

https://gist.github.com/tleyden/ea17880d3696ae64c63a#file-towards-end-of-file-L410

Towards the end of the logs, there are cases where [updateDoc() took 300+ seconds](https://gist.github.com/tleyden/ea17880d3696ae64c63a#file-towards-end-of-file-L410).  At first I thought there was an issue with instrumentation, but this does correlate with the response times reported by gateload:

```
PostBulkDocs: {
count: 17213,
sum: 2227437559208583,
min: 7679315,
max: 775360912380,  (775 seconds)
mean: 129404378040.35223,
p95: 618772927542, (618 seconds)
p99: 721646494030 (721 seconds)
},
```

Action item:
- Continue to add tracing/instrumentation to try to narrow this down

**Comment by tleyden** [comment]
Based on log messages such as these:

```
2015-05-19T15:54:58.278-07:00 updateDoc-mark_affected_users_roles() took 110.55192509 seconds
```

It appears that a lot of the write bottlenecks are coming from this code block in the `func (db *Database) updateDoc()` method:

```
    // Mark affected users/roles as needing to recompute their channel access:
    if len(changedPrincipals) > 0 {
        base.LogTo("Access", "Rev %q/%q invalidates channels of %s", docid, newRevID, changedPrincipals)
        for _, name := range changedPrincipals {
            marker, enterTime = base.TraceEnterExtra("invalid_user_or_role_channels")
            db.invalUserOrRoleChannels(name)
            base.TraceExit(marker, enterTime)
            //If this is the current in memory db.user, reload to generate updated channels
            if db.user != nil && db.user.Name() == name {
                marker, enterTime = base.TraceEnterExtra("authenticator_get_user")
                user, err := db.Authenticator().GetUser(db.user.Name())
                base.TraceExit(marker, enterTime)
                if err != nil {
                    base.Warn("Error reloading db.user[%s], channels list is out of date --> %+v", db.user.Name(), err)
                } else {
                    db.user = user
                }
            }

        }
    }
```

I'm continuing to drill down further.

**Comment by tleyden** [comment]
Based on log message:

```
2015-05-19T16:50:19.014-07:00 getPrincipal-delta_invoke_update_and_callback_invocation() took 352.167148123 seconds
```

The bottlenecks appear to be in:

```
// Common implementation of GetUser and GetRole. factory() parameter returns a new empty instance.
func (auth *Authenticator) getPrincipal(docID string, factory func() Principal) (Principal, error) {

    var princ Principal

    err := auth.bucket.Update(docID, 0, func(currentValue []byte) ([]byte, error) {

        ...
```

Where the measured delay is happening between calling Update and when the callback function is actually invoked.

I'm adding instrumentation into go-couchbase to try to drill down further.

**Comment by tleyden** [comment]
From [perf run 947](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/947/):

## Cas loop bottlenecks

I'm seeing logs like these:

```
2015/05/20 14:39:02 update-gocb-update_cas_loop() took 564.486664676 seconds
2015/05/20 14:39:02 update-gocb-update_cas_loop() numUpdateAttempts: 34
```

Which are from [this code](https://github.com/couchbase/go-couchbase/blob/master/client.go#L658-L668):

```
// internal version of Update that returns a CAS value
func (b *Bucket) update(k string, exp int, callback UpdateFunc) (newCas uint64, err error) {
    var state memcached.CASState
    for b.casNext(k, exp, &state) {
        var err error
        if state.Value, err = callback(state.Value); err != nil {
            return 0, err
        }
    }
    return state.Cas, state.Err
}
```

## Wait for persistence bottlenecks

Seeing logs such as:

```
2015/05/20 14:36:02 WriteUpdate-gocb-wait_for_persist_index() took 6.569289182 seconds
2015/05/20 14:36:02 WriteUpdate-gocb-wait_for_persist_index() took 22.167946631 seconds
2015/05/20 14:36:03 WriteUpdate-gocb-wait_for_persist_index() took 22.371605465000002 seconds
```

Which is from [this code](https://github.com/couchbase/go-couchbase/blob/master/client.go#L679-L698):

```
func (b *Bucket) WriteUpdate(k string, exp int, callback WriteUpdateFunc) error {
    ...
    // If callback asked, wait for persistence or indexability:
    if writeOpts&(Persist|Indexable) != 0 {
        err = b.WaitForPersistence(k, cas, deletion)
    }
    return err
}
```

**Comment by tleyden** [comment]
- Figure out the doc for cas loops
- In wait for persistence, is it a sleep loop or observe taking the time?
- Instrument how long things are waiting on a pool connection
- Bump the sleepDelay := 500 \* time.Millisecond  // starts highter, chews up less connections

**Comment by tleyden** [comment]
From [perf run 948](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/948/):

> Figure out the doc for cas loops

They all seem to be user docs:

```
2015/05/20 17:18:38 update-gocb-update_cas_loop-docid:_sync:user:user-3271() took 654.195775711 seconds.  docid: _sync:user:user-3271
2015/05/20 17:18:38 update-gocb-update_cas_loop-docid:_sync:user:user-3271() numUpdateAttempts: 57
```

> In wait for persistence, is it a sleep loop or observe taking the time?

A lot of time seems to be lost to sleeping.  The longest one I saw was:

```
2015/05/20 17:05:58 gocb-waitforpersistence sleeping 7.389459312s after 19 attempts so far
```

**Comment by tleyden** [comment]
I tried Sherlock build 4.0.0-2115 in [perf run 949](http://ci.sc.couchbase.com/view/sync%20gateway/job/cen006-p0-cbsgw-real-perf/949/parameters/), and I didn't see any overall improvement on the PostBulkDocs P95 stats.

However, there were no more `gocb-waitforpersistence sleeping` messages.

Still seeing large cas loop logs:

```
2015/05/20 17:41:17 update-gocb-update_cas_loop-docid:_sync:user:user-5827() took 243.680430818 seconds.  docid: _sync:user:user-5827
2015/05/20 17:41:17 update-gocb-update_cas_loop-docid:_sync:user:user-5827() numUpdateAttempts: 32
```

**Comment by tleyden** [comment]
Action item: factor into separate tickets
- [ ] Is computeChannels in the write path
- [ ] Factor out ticket: multiple docs doing access grant, clearing user (symptom: lots of cas loops on user)

**Comment by zgramana** [comment]
@tleyden please update or preferably create new tickets and close this one.

**Comment by tleyden** [comment]
- [x] Ideally we can reproduce this via gatling scenario
- [ ] Come up with a solution for high user doc contention

**Comment by tleyden** [comment]
> Ideally we can reproduce this via gatling scenario

See #920

**Comment by tleyden** [comment]
Closing this in favor of https://github.com/couchbase/sync_gateway/issues/937
